pkg/distributor/distributor.go

```diff
@@ -598,26 +609,69 @@ func (d *Distributor) validateSeries(nowt time.Time, ts mimirpb.PreallocTimeseri
 	return nil
 }
 
-// forwardingReq returns a forwarding request if one is necessary, given the user ID.
-// if no forwarding request is necessary it returns nil.
-func (d *Distributor) forwardingReq(ctx context.Context, userID string) forwarding.Request {
+// PrePushForwardingMiddleware is used as push.Func middleware in front of PushWithCleanup method.
+// It forwards time series to configured remote_write endpoints if the forwarding rules say so.
+func (d *Distributor) PrePushForwardingMiddleware(next push.Func) push.Func {
 	if d.forwarder == nil {
-		return nil
+		// Forwarding is disabled, no need to wrap "next".
+		return next
 	}
 
-	forwardingRules := d.limits.ForwardingRules(userID)
+	return func(ctx context.Context, req *mimirpb.WriteRequest, cleanup func()) (*mimirpb.WriteResponse, error) {
+		userID, err := tenant.TenantID(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		var errCh <-chan error
+		req.Timeseries, errCh = d.forwardSamples(ctx, userID, req.Timeseries)
+		resp, nextErr := next(ctx, req, cleanup)
```

Comment:
The docs on your `forwarder.Forward` method now say:

> The slice of time series which gets passed into this function must not be returned to the pool by the caller

However, this `next(..., cleanup())` call will most likely return `req.Timeseries` back to the pool, right?

Reply:
It will do here:

https://github.com/grafana/mimir/blob/4cc0bb1670a7b69912d40c4f6185a3ecd415c1ab/pkg/util/push/push.go#L84-L87

Reply:
Discussed offline: this works because we reuse the updated `req.Timestamp` here, not the one we've created.

Reply:
marking as resolved